src/librustc_codegen_ssa/mir/analyze.rs

```diff
@@ -261,9 +131,67 @@ impl<'mir, 'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> Visitor<'tcx>
         self.super_terminator(terminator, location);
     }
 
-    fn visit_place(&mut self, place: &mir::Place<'tcx>, context: PlaceContext, location: Location) {
+    fn visit_place(
+        &mut self,
+        place: &mir::Place<'tcx>,
+        mut context: PlaceContext,
+        location: Location,
+    ) {
         debug!("visit_place(place={:?}, context={:?})", place, context);
-        self.process_place(&place.as_ref(), context, location);
+
+        // Non-uses do not force locals onto the stack.
+        if !context.is_use() {
+            return;
+        }
+
+        let mir::Place { local, projection } = *place;
+
+        // Reads from ZSTs do not require memory accesses and do not count when determining what
+        // needs to live on the stack.
+        if is_consume(context) {
+            let ty = place.ty(self.fx.mir, self.fx.cx.tcx()).ty;
+            let ty = self.fx.monomorphize(&ty);
+            let span = self.fx.mir.local_decls[local].source_info.span;
+            if self.fx.cx.spanned_layout_of(ty, span).is_zst() {
+                return;
+            }
+        }
+
+        let mut has_disqualifying_projection = false;
+
+        let mut projection: &[_] = projection.as_ref();
+        while let [ref proj_base @ .., elem] = *projection {
+            projection = proj_base;
+            self.super_projection_elem(local, proj_base, elem, context, location);
+
+            // Projections like `(*x)[12]` are allowed but not `*(x[12])`, since a `Deref` of a
+            // local acts like a `Copy` of that local.
+            if let mir::PlaceElem::Deref = elem {
+                has_disqualifying_projection = false;
+                context = PlaceContext::NonMutatingUse(NonMutatingUseContext::Copy);
+                continue;
+            }
+
+            // Ignoring `Deref`s, the only allowed projections are reads of scalar fields.
+            if is_consume(context) && matches!(elem, mir::ProjectionElem::Field(..)) {
+                let base_ty = mir::Place::ty_from(local, proj_base, self.fx.mir, self.fx.cx.tcx());
+                let base_ty = self.fx.monomorphize(&base_ty);
+                let span = self.fx.mir.local_decls[local].source_info.span;
+                let layout = self.fx.cx.spanned_layout_of(base_ty.ty, span);
+
+                if !ty_requires_alloca(self.fx, layout) {
```

Comment:
Hmm, I feel like this doesn't capture the important aspect here: only for *some* layouts we can extract their fields as SSA values. So maybe instead of `!ty_requires_alloca` it could be `layout_allows_ssa` (i.e. with flipped meaning).